Rebase, Client-Certificates, Double/Split `authn` with Okta

coldcoff opened the pull request:
Hi @aclindsa,

This patch make it possible to use `gp-okta.py` with our corporate VPN.

1. rebase against changes in arthepsy:master since @aclindsas fork

2. add possibility for using a client certificate, see [Issue 17](https://github.com/arthepsy/pan-globalprotect-okta/issues/17) from @lvml.

3. Implement double `authn` login, get `stateToken` from redirects for doing an additional `authn` with Okta. (Note that we indeed need this double `authn`  *together* with your `another_dance` to connect to our corporate VPN)

Please check and probably merge with your repo, all should be backwards compatible.

Ideally this can be merged in the original @arthepsy repo with your existing [Pull Request #7](https://github.com/arthepsy/pan-globalprotect-okta/pull/7) then to fullfill @dlenski idea of a `one-globalprotect-saml-script-to-rule-them-all` script)

Thank you for all your work

Tino Lange
(coldcoff)

Commit:
Implement config option `cert` to use a client certificate.

This fixes Issue https://github.com/arthepsy/pan-globalprotect-okta/issues/17 raised by @lvml.

(Note that if you need to specify a client certificate for the `vpn_url`, then most probably
you will also need to give the same certificate to the final `openconnect` call with `--certificate` via the `openconnect_args`)

Commit:
Implement double authn login (via stateToken) needed for some corporate VPN setups.
The first authn call only returns the `sessionToken`, with the redirect later one gets the `stateToken` and afterwards need to do the authn call again for full authentication (and mfa).

aclindsa commented:
I've cherry-picked this locally and after some munging, it appears that it doesn't stop my VPN login from working, though it *does* seem like it slows down the authentication dance... is that expected?

aclindsa commented on gp-okta.conf:
```diff
@@ -8,8 +8,9 @@ sms.okta = 0
+#cert = path-to-client-cert-as-unencrypted-pem-file.pem
 #openconnect_cmd = sudo openconnect
-openconnect_args =                    # optional arguments to openconnect
+openconnect_args =                    # optional arguments to openconnect, probably you might want to repeat the cert here (if used above) with --certificate=xxx
```
Could this be added automatically by the script if`cert` is specified in the config file? If it's always required, we might as well only have one place it has to be configured...

coldcoff replied:
Good point. I also thought about that.
But probably one needs some other certificate for the openconnect call? Or only one for the "dance"? In the end I left it open, offering the most flexibility (only with the comment as hint) as it would be hard to "stop" such an automatism if the setup is different and only requires the certificate once. I think the `gp-okta.conf` is still small enough to mentally understand all of it.

aclindsa replied:
I confess I don't understand enough about your setup to understand why the two certificates are probably different. Why is that?

Anyway, would you mind building upon this comment to explain when the certificates would and would not be the same (I'm certainly confused, perhaps the next person to look at this will be, too)?

coldcoff replied:
Well, if you don't like it ... for me they are also identical -- if one needs that differently, he/she can raise a pull request lateron himself ;-)
-> Changed in fab69fe18ccc75c9e63dc1c94c96a47547aca4f4

aclindsa replied:
Sorry if it came off as such, but my last comment was not intended to goad you into doing what I'd initially requested.

I honestly don't understand how these certificates are used, and just wanted to make sure we're doing the right thing. I think that the flexibility you mentioned is a good reason to keep them separate config options as long as we expect someone to actually use it.

Are you happy with the current state of this PR, or do you still prefer them separate? I want this to be the correct decision, and not to be made based on what you think I want...

coldcoff replied:
I'm fine with the current state of this PR, simply as it works for me - in my use case there is only one certificate involved.

But: please see @lvml comment below. He made a valid point - not the application is relevant (python 'requests' vs. openconnect) - in fact the destination domain (i. e. vpn-gateway vs. okta) should be relevant.

As the https://github.com/aclindsa/pan-globalprotect-okta/blob/5f24bc5ec60013f6b2096a74124e01212b36b2de/gp-okta.py#L448 is used for vpn_url *and* okta_url altogether a separation of the certificates (if you want that) should probably be better based on the destination.

Validation is some other point, see discussion below.

aclindsa commented on gp-okta.py:
```diff
@@ -446,6 +446,9 @@ def main():
 	conf = load_conf(sys.argv[1])
 	
 	s = requests.Session()
+        if conf.get('cert'):
```
Please check your tabs vs. spaces here and in your other commit. I had to fix a few places that errored out with:
```
  File "./gp-okta.py", line 390
    if state_token is not None:
                              ^
TabError: inconsistent use of tabs and spaces in indentation
```

coldcoff replied:
Oh, you are right. I tested with python2.7 -- that one did not complain, whereas python3.x does.

Fixed.

coldcoff commented:
> I've cherry-picked this locally and after some munging, it appears that it doesn't stop my VPN login from working, though it _does_ seem like it slows down the authentication dance... is that expected?

Oh. This is interesting. Does it change your "dance"?
Do you now make the `authn` call also twice? If yes, you probably get a `stateToken` in the redirect Javascript as well and the new code reacts on that although in your case this is not needed as your first `authn` was sufficient? Could you please check that? If this is the case we can add some bool to additionally decide on the second `authn` call.

Commit:
Fix: inconsistent spacing (tabs<->spaces)

Commit:
Set certificate automatically also in openconnect call, if one is set in `gp-okta.conf` for the requests-dance

lvml commented:
While nobody as of yet has documented a use case for presenting different
client certificates to the GlobalProtect VPN gateway and the domain.okta.com
service, I think such use cases may well exist.

Okta seems to support the use of x.509 certificate for "device identification":
https://help.okta.com/en/prod/Content/Topics/Mobile/Okta_Mobile_Device_Trust_Windows-desktop.htm

We already know of the GlobalProtect VPN gateways that they support
the use of x.509 certificates for client authentication, and these
two uses could involve different certificates issued by different CAs.

But there is another, even more compelling reason why to use different
configuration items for the x.509 certificates of GlobalProtect vs. Okta:
In order to not be fooled by a fake VPN gateway or a fake Okta authentication
service that could act as man-in-the-middle adversaries, the client
needs to check that the server certificate is trusted to him.
Thus the configuration items are not only relevant for the client certificate,
but also for providing the CA trust chain that the client can use to
verify the server certificate.

(Or, instead of CA trust chain validation, which is always prone
to attacking via not-really-trustworthy CAs, the client could
have a configuration item with the actual public key of the VPN gateway
and Okta service to use. The client would need to update those,
of course, if the public key of the servers change.)

coldcoff commented:
@lvml, you are right.

But as https://github.com/aclindsa/pan-globalprotect-okta/blob/5f24bc5ec60013f6b2096a74124e01212b36b2de/gp-okta.py#L227 will always use a fixed URL and a fixed destination host (even if the vpn_url is a fake VPN gateway) the attacker has at least no chance (by malicious redirects for example) to let `gp-okta.py` POST the credentials to some other destination.

Nevertheless I think we should probably add some certificate checking, as you recommended. Additionally  I just see that `openconnect` itself does implement such a certificate checking already: https://github.com/openconnect/openconnect/blob/0120a25b4584ec231d48f9aaa15580e1dc5ac435/main.c#L824. This will address the case that `okta_url` is a fake and malicious.

coldcoff commented:
@aclindsa :

> > ... it _does_ seem like it slows down the authentication dance... is that expected? 

> Oh. This is interesting. Does it change your "dance"?
> Do you now make the `authn` call also twice? ... Could you please check that? ...

Did you already have a chance to doublecheck that?

coldcoff commented:
(I'm preparing a set of patches, stay tuned)

coldcoff committed:
- Add ssl cert verification for vpn and okta side, if configured
  - new option `vpn_url_cert`
  - new option `okta_url_cert`
- rename client certificate option to `client_cert` from `cert`
- totp: comment out ABCDEFGHIJKLMNOP secrets, default to not set (so you get asked)
- grab and display prelogin errors (for example: client certificate needed or such)
- new option `openconnect_certs` that points to a file to collect all given and collected server certificates, will be given to the final openconnect call (inspired by nicklans fork, thank you). If it is not set it will be a temp file.
- check all redirect and parsed urls if they point to an expected domain (either `okta_url` or `vpn_url`)
- learn `gateway` from getconfig call, if not set/overridden by config file

coldcoff commented:
OK, I just commited some changes that sould address the verification issues while staying backwards compatible to the users not needing client certificates. Nevertheless they can now configure verification as well. Please test.

coldcoff committed:
- flush openconnect_certs file in between and finally close before giving to openconnect call

coldcoff committed:
do not close file, flush is enough for now - otherwise it might get deleted in case of the temp file solution

coldcoff committed:
Use stderr for error messages.

coldcoff committed:
Shorten imports

coldcoff committed:
Add the committing people all to the copyright notice

coldcoff committed:
Do the "another_dance" at the desired gateway to obtain a cookie that is valid there

coldcoff commented:
@aclindsa please test my recent changes.
The reason for the "another_dance" is that probably you change the machine from portal to gateway.

Imagine a portal that shows you gateways all over the world, and you choose one nearest to your location.

The second dance is needed to get a proper cookie, doing it at the destination gateway is important as the prelogin_cookie from one gateway is not valid on another.

I asume that in your setup the gateway and the portal were the same, that is why you could (and needed to) direct your second dance to the very same machine (vpn_url).

Please test -- as far as I can see (and hope :-) ) I did not break anything for you. For us the new version allows to choose one of the many gateways that our company offers.

aclindsa commented:
I confess I haven't looked through all the code yet, but I tried to check out and use your latest and got:

```
[INFO] getconfig request [vpn_url]
Traceback (most recent call last):
  File "/home/aclindsa/bin/.pan-globalprotect-okta/gp-okta.py", line 588, in <module>
    main()
  File "/home/aclindsa/bin/.pan-globalprotect-okta/gp-okta.py", line 529, in main
    userauthcookie, gateway = paloalto_getconfig(conf, s, saml_username, prelogin_cookie)
  File "/home/aclindsa/bin/.pan-globalprotect-okta/gp-okta.py", line 481, in paloalto_getconfig
    for entry in x.find('.//root-ca'):
TypeError: 'NoneType' object is not iterable
```

Is this another python2 vs python3 thing?

coldcoff committed:
Some portals seem not to return root-ca entries, detect that and just continue

coldcoff commented:
@aclindsa, please retry.

coldcoff commented:
@aclindsa - for us here the dust has settled. this `gp-okta.py` works day-to-day for me and many colleagues.

Were you able to re-test and have a look at it meanwhile?

Once you merged my changes from https://github.com/coldcoff/pan-globalprotect-okta to your https://github.com/aclindsa/pan-globalprotect-okta we can afterwards ask @arthepsy to merge your (then updated) pull request https://github.com/arthepsy/pan-globalprotect-okta/pull/7. Then everything is back home.

aclindsa commented:
@coldcoff Wow, you've done a lot of work! So much, in fact, that it's grown outside the scope of what can be a simple tag-along to [my PR](https://github.com/arthepsy/pan-globalprotect-okta/pull/7).

I *have* now tested your latest code and confirm that I can still connect to my VPN, but I don't understand your changes well enough to be able to address any questions/concerns @arthepsy might have. I suggest you open your own PR against his repository to take this farther, since I also don't have the time or interest to maintain a permanent fork.

coldcoff committed:
Bugfix: guard case when error msg is not returned

coldcoff committed:
Bugfix: use unicode literals also on python2, this avoids errors with formatstring.format(...)
(see: https://stackoverflow.com/questions/3235386/python-using-format-on-a-unicode-escaped-string)

coldcoff committed:
Feature: Allow Yubikey `webauthn` authentication, if configured as 2FA in Okta.
Just insert your YubiKey and press the blinking button, when asked to do so.
Needs `fido2` packages to work (`pip install fido2`).

coldcoff commented:
... now I have also added webauthn support (this is to use a YubiKey instead of typing in a TOTP secret).

OK, @aclindsa -- I'll open a separate PR against @arthepsy.
All your changes are on biard as well, as I initially had forked from your master, so if I get @arthepsy to merge this will implicitely merge your PR also.

coldcoff commented:
New, combined PR is: https://github.com/arthepsy/pan-globalprotect-okta/pull/19
